src/prommis/nanofiltration/costing/diafiltration.py

```diff
@@ -0,0 +1,713 @@
+    assert not hasattr(
+        m.fs.stage3, "permeate_energy_balance"
+    )  # check that there are no energy balances
+    assert not hasattr(
+        m.fs.stage3, "permeate_pressure_balance"
+    )  # check that there are no pressure balances
+
+    m.fs.stage3.length = pyo.Var(units=pyo.units.m)
+    m.fs.stage3.length.fix(10)
+
+    m.fs.stage3.solvent_flux = pyo.Constraint(
+        m.fs.stage3.elements,
+        rule=solvent_rule,
+    )
+    m.fs.stage3.solute_sieving = pyo.Constraint(
+        m.fs.stage3.elements,
+        m.fs.solutes,
+        rule=stage3_solute_rule,
+    )
+
+    # Add second feed mixer
+    m.fs.mix2 = Mixer(
+        num_inlets=2,
+        property_package=m.fs.properties,
+        material_balance_type=MaterialBalanceType.componentTotal,
+        energy_mixing_type=MixingType.none,
+        momentum_mixing_type=MomentumMixingType.none,
+    )
+
+    # Add 3 Arcs to connect third stage to remainder of flowsheet
+    m.fs.stream4 = Arc(
+        source=m.fs.stage2.permeate_outlet,
+        destination=m.fs.mix2.inlet_2,
+    )
+    m.fs.stream5 = Arc(
+        source=m.fs.mix2.outlet,
+        destination=m.fs.stage3.retentate_inlet,
+    )
+    m.fs.stream6 = Arc(
+        source=m.fs.stage3.retentate_outlet,
+        destination=m.fs.mix1.inlet_1,
+    )
+
+    pyo.TransformationFactory("network.expand_arcs").apply_to(m)
+
+    # Unfix Feed to Mixer 1 inlet 1
+    m.fs.mix1.inlet_1.flow_vol[0].unfix()
+    m.fs.mix1.inlet_1.conc_mass_solute[0, "Li"].unfix()
+    m.fs.mix1.inlet_1.conc_mass_solute[0, "Co"].unfix()
+
+    # Fix Diafiltrate feed to Mixer 2 inlet 1
+    m.fs.mix2.inlet_1.flow_vol[0].fix(30)
+    m.fs.mix2.inlet_1.conc_mass_solute[0, "Li"].fix(0.1)
+    m.fs.mix2.inlet_1.conc_mass_solute[0, "Co"].fix(0.2)
+
+    # Fix Feed stream at element 10 of stage 3
+    m.fs.stage3.retentate_side_stream_state[0, 10].flow_vol.fix(Q_feed)
+    m.fs.stage3.retentate_side_stream_state[0, 10].conc_mass_solute["Li"].fix(C_Li_feed)
+    m.fs.stage3.retentate_side_stream_state[0, 10].conc_mass_solute["Co"].fix(C_Co_feed)
+
+    assert degrees_of_freedom(m) == 0
+
+    # Initialize Mixer 2
+    propagate_state(
+        destination=m.fs.mix2.inlet_2,
+        source=m.fs.stage2.permeate_outlet,
+    )
+
+    mix_initializer.initialize(m.fs.mix2)
+
+    # Try to initialize stage 3
+    propagate_state(
+        destination=m.fs.stage3.retentate_inlet,
+        source=m.fs.mix2.outlet,
+    )
+
+    initializer.initialize(m.fs.stage3)
+
+    # Lets try again
+    # Initial feed guess is pure diafiltrate (no recycle)
+    # Remember to only set the value, not fix the inlet
+    m.fs.stage3.retentate_inlet.flow_vol[0].set_value(Q_diaf)
+    m.fs.stage3.retentate_inlet.conc_mass_solute[0, "Li"].set_value(C_Li_diaf)
+    m.fs.stage3.retentate_inlet.conc_mass_solute[0, "Co"].set_value(C_Co_diaf)
+
+    initializer.initialize(m.fs.stage3)
+
+    add_expressions(m)
+
+    return m
+
+
+def create_rules(m):
```

Comment:
Not necessarily wrong, but it seems odd to me to have a function which creates the functions/rules for the `Constraints`.

Firstly, the `@model.Constraint` approach to creating constraints (and other components) is generally more compact and easier to read.
Secondly, why not just create these at the scope where they are used?